acracensor/handlers/querycapture_handler.go

```diff
@@ -0,0 +1,216 @@
+package handlers
+
+import (
+	"strings"
+	"io/ioutil"
+	"encoding/json"
+	"os"
+	"github.com/cossacklabs/acra/logging"
+	log "github.com/sirupsen/logrus"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+const MaxQueriesInChannel = 10
+const DefaultSerializationTimeout = time.Second
+
+type QueryCaptureHandler struct {
+	Queries              []QueryInfo
+	filePath             string
+	logChannel           chan QueryInfo
+	signalToSerialize    chan bool
+	signalBackgroundExit chan bool
+	serializationTimeout time.Duration
+}
+
+type QueryInfo struct {
+	RawQuery string
+	IsForbidden bool
+}
+
+func NewQueryCaptureHandler(filePath string) (*QueryCaptureHandler, error) {
+	var _, err = os.Stat(filePath)
+
+	// create file if not exists
+	if err != nil {
+		if os.IsNotExist(err) {
+			var file, err = os.Create(filePath)
+			if err != nil {
+				return nil, err
+			}
+			defer file.Close()
+		} else {
+			return nil, err
+		}
+	}
+
+	bufferBytes, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	var queries []QueryInfo
+
+	if len(bufferBytes) != 0 {
+		if err = json.Unmarshal(bufferBytes, &queries); err != nil{
+			return nil, err
+		}
+	}
+
+	logChannel := make(chan QueryInfo, MaxQueriesInChannel)
+	signalToSerialize := make(chan bool)
+	signalShutdown := make(chan os.Signal, 2)
+	signal.Notify(signalShutdown, os.Interrupt, syscall.SIGTERM)
+
+	signalBackgroundExit := make(chan bool)
+
+	handler := &QueryCaptureHandler{}
+	handler.filePath = filePath
+	handler.Queries = queries
+	handler.signalToSerialize = signalToSerialize
+	handler.logChannel = logChannel
+	handler.signalBackgroundExit = signalBackgroundExit
+	handler.serializationTimeout = DefaultSerializationTimeout
+
+	//serialization signaling goroutine
+	go func() {
+		for {
+			select {
+			case <-signalBackgroundExit:
+				return
+			case <-signalShutdown:
+				return
+			default:
+				//do nothing. This means that channel has no data to read yet
+			}
+			time.Sleep(handler.serializationTimeout)
+			//timer finished. Close channel to inform that serialization should be performed
+			signalToSerialize <- true
+		}
+	}()
+
+	//handling goroutine
+	go func (){
+		for {
+			select {
+			case <-signalToSerialize:
```

Comment:
I tried as you proposed (`<- time.NewTimer(...).C`) but it didn't work as expected. Test ran infinitely, so I left this unchanged

Reply:
it will works as work now. your goroutine above do the same - infinitely send signal to `signalToSerialize` and stop when took `signalShutdown`
When you will use `NewTimer`, you will infinitely receive signal to serialize at constant interval and will stop on `signalShutdown` (this signal we handle here too)
https://play.golang.org/p/l8uYMdPePbC - example
This is last change for it PR that we need and will simplify processing. Replace signalToSerialize to Ticker